Base Implementation of `LibPrices`

sporejack opened the pull request:
# Motivation
Cleaner pricing logic and types (the most important of all!).

# Changes
 - Add `FundingRateInstant` type
 - Remove `Types.FundingRate` type
 - Fix contracts to use new funding rate type

sporejack committed:
Add `LibPrices` library

This commit defines the `LibPrices` Solidity library.

sporejack committed:
Correct name of funding rate type

sporejack committed:
Update `IPricing` interface to use `LibPrices`

This commit modifies the `IPricing` interface to use the new funding
rate type (`FundingRateInstant`) defined in the `LibPrices` library.

sporejack committed:
Remove old funding rate type

sporejack committed:
Patch contracts to use new funding rate type

sporejack committed:
Fix formatting

raymogg commented on contracts/Pricing.sol:
```diff
@@ -268,43 +257,34 @@ contract Pricing is IPricing {
-            fundingRate.fundingRate,
-            fundingRate.fundingRateValue
-        );
+        Prices.FundingRateInstant memory fundingRate = fundingRates[index];
```
Why create a new FundingRateInstant here? Should just return the fundingRate param in memory

raymogg commented on contracts/Pricing.sol:
```diff
@@ -268,43 +257,34 @@ contract Pricing is IPricing {
-            fundingRate.fundingRateValue
-        );
+        Prices.FundingRateInstant memory fundingRate =
+            insuranceFundingRates[index];
```
Same as above. No need to create a whole new FundingRateInstant struct

sporejack replied:
Consult e317c30.

sporejack committed:
Add `globalLeverage` function to `LibPrices`

This commit adds the `globalLeverage` function to the `LibPrices`
library. This function calculates the new global leveraged notional
value from old and new account states.

sporejack committed:
Define price instant type

sporejack committed:
Patch types file to use new pricing type

sporejack committed:
Add TWAP type to LibPrices

sporejack committed:
Add `calculateTWAP` to `LibPrices`

This commit implements TWAP calculation in `LibPrices`.

sporejack committed:
Add array utilities to maths library

sporejack committed:
Implement average pricing for arbitrary periods

sporejack committed:
Fix array types in parameters

sporejack committed:
Fix array types in LibPrices

sporejack committed:
Patch other contracts to use new pricing libs

sporejack committed:
Use new pricing lib for 24 hour prices

sporejack committed:
Extract fair pricing to pricing library

sporejack committed:
Extract time value into LibPrices

sporejack committed:
Simplify getters in pricing contract

sporejack committed:
Merge branch 'contracts-v2' into 271-base-for-libprices

sporejack committed:
Fix formatting

CalabashSquash commented on contracts/Interfaces/Types.sol:
```diff
@@ -2,23 +2,18 @@
-    struct PricingMetrics {
-        Types.HourlyPrices[24] hourlyTracerPrices;
-        Types.HourlyPrices[24] hourlyOraclePrices;
+    struct PricesMetrics {
```
Nitpick: Why is `PriceMetrics` in here and not in `LibPricing.sol`?

CalabashSquash replied:
Also, is `struct PriceMetrics` even being used anymore?

sporejack replied:
Shouldn't even exist; good catch.

sporejack replied:
Consult cb37d4f

sporejack committed:
Prune old types file

raymogg commented on contracts/TracerPerpetualSwaps.sol:
```diff
@@ -359,7 +332,7 @@ contract TracerPerpetualSwaps is
      *      and the account must then be liquidated.
      * @param account the address to settle.
-     * @dev This function aggregates data to feed into account.sol"s settle function which sets
+     * @dev This function aggregates data to feed into account.sols settle function which sets
```
Comment needs changing. not part of this PR but just noting

raymogg commented on contracts/lib/LibMath.sol:
```diff
@@ -12,4 +12,21 @@ library LibMath {
     function abs(int256 x) internal pure returns (int256) {
         return x > 0 ? int256(x) : int256(-1 * x);
     }
+
```
I am fairly certain the PRB math library has quite efficient implementation of these. Also since fundingRate will most likely end up being a WAD value (since the price data its based off is WAD values) we will want to eventually use that

raymogg replied:
actually re thinking, the fact that they are WAD values shouldn't matter. Its more a question of which implementation is more efficient

sporejack replied:
@raymogg Honestly all for a non-homemade solution for these -- I'd prefer to not have to write them myself. I agree about the correctness not being impacted and that we should then prioritise efficiency.

sporejack committed:
Fix formatting!

raymogg commented on contracts/lib/LibPrices.sol:
```diff
@@ -0,0 +1,121 @@
+            } else {
+                return
+                    TWAP(
+                        cumulativeUnderlying / instantUnderlying,
```
These should be WAD divides since they are simply a scaled underlying average price, which is a WAD value. 

If we don't use WAD division, the returned value will not be a WAD and will throw out a lot.

sporejack replied:
Consult 7d054b5.

CalabashSquash commented on contracts/Pricing.sol:
```diff
@@ -217,46 +214,38 @@ contract Pricing is IPricing {
+     * @param cumulativeFundingRate The cumulativeFundingRate, incremented each time the funding rate is updated
      */
-    function setFundingRate(
-        uint256 marketPrice,
-        int256 fundingRate,
-        int256 fundingRateValue
-    ) internal {
-        fundingRates[currentFundingIndex] = Types.FundingRate(
+    function setFundingRate(int256 fundingRate, int256 cumulativeFundingRate)
+        internal
+    {
+        fundingRates[currentFundingIndex] = Prices.FundingRateInstant(
             block.timestamp,
-            marketPrice,
             fundingRate,
-            fundingRateValue
+            cumulativeFundingRate
         );
```
So just to clarify, is the reason for removing the `marketPrice` parameter just because its not really necessary when you'r ekeeping track of funding rates at different points in time?

sporejack replied:
`marketPrice` is simply a redundant field -- we came to the conclusion it was always redundant (even in its previous usage).

CalabashSquash commented on contracts/lib/LibPrices.sol:
```diff
@@ -0,0 +1,121 @@
+        if (oldLeverage < newLeverage) {
+            return globalLeverage + delta;
+        } else {
+            if (delta == globalLeverage) {
+                return 0;
+            } else {
```
I'm not sure I understand this. If the change == the global leveraged value, the global leveraged value becomes 0?

What if delta == 100, globalLeverage == 100, shouldn't globalLeverage become 200?

sporejack replied:
So, in the updated version of this function body:

```solidity
int256 delta = int256(newLeverage - oldLeverage);

if (delta >= 0) { /* leverage has increased or remained the same */
    return globalLeverage + uint256(delta);
} else { /* leverage has decreased */
    if (uint256(delta) > globalLeverage) { /* handle underflow */
        return 0;
    } else {
        return globalLeverage - uint256(delta);
    }
}
```

This implements a floor for `globalLeverage`, due to it being unsigned.

CalabashSquash replied:
I think this would still not work, because:
- (discussed in person:) int256(newLeverage - oldLeverage) would still underflow before being case to int256
- in the `else` block, you cast `delta` to `uint256`, but `delta` has to be <0 to enter this branch, leading to an invalid cast.

CalabashSquash commented on contracts/lib/LibPrices.sol:
```diff
@@ -0,0 +1,121 @@
+    ) public pure returns (uint256) {
+        uint256 delta = oldLeverage - newLeverage;
+
+        if (oldLeverage < newLeverage) {
+            return globalLeverage + delta;
+        } else {
+            if (delta == globalLeverage) {
+                return 0;
+            } else {
```
Consider this e.g.
newLeverage = 200
oldLeverage = 100
then delta = -100 (but delta is unsigned so underflow)

Is this a concern?

sporejack replied:
> Is this a concern?

Yes, I believe the newer version of this block addresses this issue. Keen to hear your input.

CalabashSquash reviewed (changes requested):
:seychelles:

sporejack committed:
WAD LibPrices

sporejack committed:
Fix arithmetic signedness issues in calculating global leverage

sporejack committed:
Fix formatting

sporejack committed:
Fix casting of delta subtraction

sporejack committed:
Replace negative delta with reversed expression

raymogg commented on contracts/lib/LibPrices.sol:
```diff
@@ -79,7 +79,7 @@ library Prices {
                 return 0;
             } else {
-                return globalLeverage - uint256(delta);
+                return globalLeverage - (oldLeverage - newLeverage);
```
Why do this? Also this changes the logic right, since `delta = newLeverage - oldLeverage`, so previously this was `globalLeverage - (newLeverage - oldLeverage)`

sporejack committed:
Refactor global leverage calculation

CalabashSquash commented on contracts/lib/LibPrices.sol:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    function calculateTWAP(
+        uint256 hour,
+        PriceInstant[] memory tracerPrices,
+        PriceInstant[] memory oraclePrices
+    ) public pure returns (TWAP memory) {
+        uint256 instantDerivative = 0;
+        uint256 cumulativeDerivative = 0;
+        uint256 instantUnderlying = 0;
+        uint256 cumulativeUnderlying = 0;
+
+        for (uint256 i = 0; i < 8; i++) {
+            uint256 currTimeWeight = 8 - i;
+            uint256 j = 8 - i;
```
Did we come to a conclusion for how this should be fixed (TWAP not taking into account the `hour`)?

sporejack replied:
@BenjaminPatch can either fold it into this PR or file a separate one (given there's a separate ticket).

raymogg commented on contracts/lib/LibPrices.sol:
```diff
@@ -68,19 +68,15 @@ library Prices {
         uint256 newLeverage
     ) public pure returns (uint256) {
-        int256 delta = int256(newLeverage) - int256(oldLeverage);
+        bool leverageHasIncreased = newLeverage > oldLeverage;
```
can probably be simplified I think. Should only need one bool it seems, as the only edge case is where `newLeverage = oldLeverage` in which case `newLeveraged - oldLeverage = 0` anyway.

Current way is more explicit, but I think you should just be able to do 

`if (leverageHasIncreased) { add to leverage } else { decrease leverage }`

sporejack committed:
Simplify global leverage logic even further
